File: product_sum/product_sum.go
package main

import "fmt"

func productSum(array interface{}, depth int) int {
    sum := 0

    switch v := array.(type) {
    case []interface{}:
        for _, item := range v {
            if nestedArray, ok := item.([]interface{}); ok {
                sum += productSum(nestedArray, depth + 1)
            } else {
                sum += item.(int)
            }
        }
    }

    return sum * depth
}

func main() {
    array := []interface{}{1, []interface{}{2, 1}}
    fmt.Println(productSum(array, 1)) 
}

File: product_sum/product_sum.ts
function productSum(array: (number | (number | any[])[])[], depth: number = 1): number {
    let sum = 0;
    
    for (const item of array) {
        if (Array.isArray(item)) {
            sum += productSum(item, depth + 1);
        } else {
            sum += item;
        }
    }

    return sum * depth;
}

const array = [1, [2, 1]];

console.log(productSum(array, 1)); 
